cmake build improvements v2

wavexx opened the pull request:
Further progress on the cmake build infrastructure

- Reformat CMakeList.txt through cmake-format due to mixed formatting
- Define, but not build single-language targets by default (these are for debugging only). The target ALL_ENGLISH and ALL_FIRMWARE still drag those in.
- Fix multi-language builds: correctly copy/patch the base binary as produced by `lang-build.py`.
- Uniform binary structure in dual-language builds.
- Fix rules not to rebuild binaries if nothing changed

The multi-language builds are still slightly larger than expected, but this will be investigated later.

wavexx committed:
cmake: Reformat to fix tab/space mixture

wavexx committed:
cmake: Define all targets, but only build ALL_MULTILANG by default

ALL_ENGLISH is mostly intended for debugging purposes

wavexx committed:
cmake: Cleanup XFLASH multi-language rules

wavexx committed:
cmake: Update the variant configuration correctly when changed

wavexx committed:
cmake: Generate .hex files only where needed

This avoids generating useless intermediate .hex files during
multi-language builds.

wavexx committed:
cmake: Correctly patch binary during multi-language builds

lang-map.py expects to use the BIN file as an input-output argument.
The resulting BIN file contains an updated _PRI_LANG_SIGNATURE symbol.

wavexx committed:
cmake: Move a few variables closer to usage

wavexx committed:
cmake: Fix/uniform non-xflash builds

wavexx committed:
cmake: Use/include Buddy formatting rules for consistency

wavexx committed:
cmake: Improve a few comments

vintagepc commented on Firmware/config.h:
```diff
@@ -62,7 +62,7 @@
 
 // This is set by the cmake build to be able to take control of
 // the language flag, without breaking existing build mechanisms.
-#ifndef CMAKE_LANG_CONTROL
```
The reason I did this was to make it blatantly obvious where the language control comes from (cmake). With this change I can see users getting frustrated/confused why their setting is not taking effect if they change it here and try to do a cmake build.

wavexx replied:
I dropped the comment.

wavexx replied:
I meant commit here.

vintagepc commented on CMakeLists.txt:
```diff
@@ -2,92 +2,114 @@ cmake_minimum_required(VERSION 3.15)
+set(MAIN_LANGUAGES
+    cs de es fr it pl
+    CACHE STRING "The list of 'main' languages to be included, in the correct order"
+    )
```
Can we agree the auto-formatter is garbage? why is one separate lines and one not?

wavexx replied:
Yeah, the formatter here is just using the argument _count_ as a threshold...

vintagepc replied:
that's why I don't agree with the principle of proliferating garbage ;) the more it spreads, the more widely accepted it becomes.

vintagepc commented on CMakeLists.txt:
```diff
@@ -2,92 +2,114 @@ cmake_minimum_required(VERSION 3.15)
 
-set (LANG_BIN_MAX 249856) # Ditto, this in xflash_layout.h but needs invocation of the preprocessor... :-/
+set(LANG_BIN_MAX 249856) # Ditto, this in xflash_layout.h but needs invocation of the
+                         # preprocessor... :-/
```
this is also garbage reformat,, why split the comment? We have large displays now, column limits are arbitrary anachronisms that need to die

wavexx replied:
cmake-format is bad, or at least the defaults we are using are pretty bad. I've set the same values as the Buddy FW for the moment and just accepted the garbage...

vintagepc replied:
Not sure I agree with that mindset, but :shrug:

wavexx replied:
Following 4 projects with where this is being used, I'm just trying to make it uniform. I hate all autoformatters in general BTW.

If you can find some better settings for cmake-format (0.6.13, not the outdated 0.6.0 as used by buddy) I'm all ears. I'll try to push it though-out. I tried a bit in the past, but it always looked like garbage no matter what I tried.

vintagepc replied:
skimmed the docs and didn't see anything particularly useful configuration wise :(

wavexx replied:
Looking at buddy, where there isn't much discipline, I came to appreciate autoformatters. But I'm with you here, cmake-format is just subpar in everything it does. Some flags help with one case and destroy 20 more. I'm just trying to jump a little bit ahead and clone what we do on MMU here.

I wish I had alternatives.

vintagepc replied:
change it back to tabs and I'll be happy /s :rofl:

arekm replied:
"ColumnLimit" setting for that one thing (?).

OctoPrint uses pre-commit hook (and tool with the same name) for checking and reformatting (and this is also wired into github actions for all commits and pull requests). 

https://pre-commit.com/
https://github.com/OctoPrint/OctoPrint/blob/maintenance/.pre-commit-config.yaml
https://github.com/OctoPrint/OctoPrint/blob/master/.github/workflows/build.yml

There are various reformatting hooks supported, including simple things like trailing-whitespace or end-of-file-fixer while also more complicated things. 

~Full list:
https://pre-commit.com/hooks.html

c++ ones are also there but can't tell if these will fit your needs.

Example, manual run, for python project:
```
(venv) pi@prusa-octopi:~/devel/OctoPrint-Sdwire $ pre-commit run --hook-stage manual --all-files
fix end of files.........................................................Passed
trim trailing whitespace.................................................Failed
- hook id: trailing-whitespace
- exit code: 1
- files were modified by this hook

Fixing octoprint_sdwire/__init__.py

check for case conflicts.................................................Passed
check json...........................................(no files to check)Skipped
check yaml...............................................................Passed
check toml...........................................(no files to check)Skipped
check for merge conflicts................................................Passed
fix python encoding pragma...............................................Passed
OctoPrint codemod: batch.................................................Passed
isort....................................................................Passed
black....................................................................Passed
flake8...................................................................Passed
prettier.................................................................Passed
eslint...................................................................Passed
```

wavexx replied:
@arekm See what we do here https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/master/.pre-commit-config.yaml

Historically though this project was always buildable with just Arduino which makes is very approachable, but is also why many things are clunky.

vintagepc commented on CMakeLists.txt:
```diff
@@ -208,298 +229,321 @@ set(AVR_SOURCES
+      #[[
+      # Check po file:
+      #COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/lang/lang-check.py --no-warning --map ${LANG_MAP} ${PO_FILE}
+      #]]
```
This should probably be re-enabled for the language builds... no?
I turned it off because it wasn't working for me but it is used by PF-build.

wavexx replied:
Yup. But this (and the following) were commented before and remained commented as I didn't have time to check them.

In this case, lang-check will emit huge amounts of output with the current build layout due to the number of targets. IMHO it needs some sort of special handling.

vintagepc commented on CMakeLists.txt:
```diff
@@ -208,298 +229,321 @@ set(AVR_SOURCES
+      COMMENT "Merging language catalogs"
+      )
+    #[[
+    #add_custom_command(OUTPUT ${LANG_FINAL_BIN}
```
if this is vestigial code, remove it.

wavexx replied:
As above. This was just a pass-through in this PR.

vintagepc reviewed (changes requested):
Needs work. Doing a build followed by a "clean all" does not clean the MK2.5 hex files from the build/ dir. 

There may also be a pathing issue with one of the targets because it leaves crap like this in the build dir too:

![image](https://user-images.githubusercontent.com/53943260/193693080-1d12dc98-0477-4051-933d-b2c81df43f2a.png)

vintagepc commented on .cmake-format.py:
```diff
@@ -0,0 +1,21 @@
+
+# If true, the parsers may infer whether or not an argument list is sortable
+# (without annotation).
+autosort = True
```
can this be dangerous for the language lists?

wavexx replied:
@3d-gussner probably knows.

I think the language list was always hand-sorted. The important part here is that we keep the two groups sorted separately, which is already the case.

wavexx committed:
cmake: Cleanup the final dual-language hex

List the file as a byproduct of the rule for correct cleanup.

wavexx commented:
I fixed the cleanup.

The path issue seems pre-existing, probably a side-effect of the nested project/s?
__ stands for dir-up here. If you look at the generated rules, all the path names are constructed as project-relative, so they include __ and these are just the paths holding the final objects.

wavexx commented:
BTW I wouldn't mind a different cmake formatter entirely, but I couldn't find alternatives.

wavexx committed:
cmake: Hard-link final hex files to reduce space overhead

cmake 3.19 introduced support for hardlinks, so use them instead of
making a copy.

vintagepc commented:
> I fixed the cleanup.
> 
> The path issue seems pre-existing, probably a side-effect of the nested project/s? __ stands for dir-up here. If you look at the generated rules, all the path names are constructed as project-relative, so they include __ and these are just the paths holding the final objects.

OK, they must be part of the generated changes then - I can't recall seeing them before but it has been a while....

Do hard-link work on windows out of the box? I can't find any clear documentation on how that works on non posix platforms...
If not that change won't work. Last I checked you needed some weird developer mode in :window: s to be able to create symlinks - it's not a stock feature.

wavexx commented:
Reading the docs, yes it does, and I guess it will lead to less issues than people copying the final binary compared to a symlink.
The only limit is that of course both paths need to be on the same filesystem, which I think is always going to be the case here unless some crazy folk decides to split the tree by hand.

wavexx commented:
@leptun can you confirm this works as expected on windows?

vintagepc reviewed (approved):
Cleans properly now - Approving provisionally once we confirm the linking behaviour.

wavexx committed:
build: Size optimization

Explicitly force-noinline xyzcal_scan_pixels_32x32_Zhop to match the
previous LTO build behavior.

wavexx committed:
cmake: Normalize incoming dependency paths

3d-gussner commented:
I have found a "bug" if `Configuration_prusa.h` is present:
- Cmake MK3S-EINSy10a_Multilang--Utility
- PF-build.sh -p1 MK3S
  - Doesn't delete `Configuration_prusa.h `
- Cmake MK3-EINSy10a_Multilang--Utility
Then both are identical.

Minor ToDos:
- The build time isn't shown on LCD Support screen just `1970-01-01`
- Project version also doesn't read the FW_FLAVOR and FW_FLAVERSION
- Also the `FW_COMMIT_NR` isn't updated
- NOT cmake related.
  - The `Update language ...` during language change flickers with FW version.

wavexx committed:
cmake: Fix existing comment

vintagepc commented on CMakeLists.txt:
```diff
@@ -315,9 +315,18 @@ add_custom_target(ALL_MULTILANG)
 function(add_base_binary variant_name)
   add_executable(${variant_name} ${FW_SOURCES} ${FW_HEADERS} ${VARIANT_CFG_DST})
   set_target_properties(${variant_name} PROPERTIES EXCLUDE_FROM_ALL TRUE)
+  # add_dependencies(${variant_name} test_wrong_config) # Uncomment to check at build time
```
We should probably enable this - otherwise someone could configure with cmake, (passing the configure-time check) and then run pf build and break the build without knowing why.

wavexx replied:
I commented below, but this results in a rule that runs even if the build should be a no-op. IMHO this should _never_ happen and sucks big time.

I couldn't find a way to do better than that via cmake though.

vintagepc replied:
There's probably a way to prevent this by specifying a dummy output/byproducts file for cmake to look at so that it won't re-run unless other dependencies have changed. Maybe there's even a way to key it off the presence of the configuration_prusa.h file :thinking:

vintagepc replied:
Thought: a little clunky but perhaps it could be e.g. a post-build command for the elf - sure, it'll still build the incorrect binary, and throw an error after that. But this way if the elf hasn't changed, it won't rerun the check either leaving it a true no-op.

wavexx replied:
It also doesn't work the way I would like, as it runs late. This target would need to be a dependency on every object.

wavexx replied:
Using a target with a dummy would be better. If you attach a post-build command it's going to run for all the variants. At least with a single global target you get just one run.

wavexx commented:
The date/time/commit will be dealt with in another PR as it needs extra code to fetch the commit information.

I now also check if a stale Configuration_prusa.h is present in the source tree at configuration time (that is, when you run cmake) and refuse to continue until it's removed.

This cannot check if the file is added _later_ after configuration. For this purpose, I also added (but left commented) a build-time check that would catch such a case. This requires one rule to run at every make/ninja invocation even for a no-op build, which is something I deeply dislike.

wavexx commented:
I wonder if we're just trying to fix the wrong thing. We could actually specify the file to include at build time with something hacky like this:

```
#ifndef VARIANT
#define VARIANT Configuration_prusa.h
#endif
#define INCLUDE(v) <v>
#include INCLUDE(VARIANT)
```

then pass `-DVARIANT=variant/path`, which would also avoid the copy. We could change current references that include "Configuration_prusa.h" directly to a small wrapper. In that case, Configuration_prusa.h would still be used as a fallback, but only when the variant header is not supplied. If that still exists in the build tree it would become irrelevant for the cmake build.

Thoughts?

wavexx committed:
build: Add Configuration_var.h as a configuration wrapper

Instead of including Configuration_prusa.h directly, include
Configuration_var which then includes the proper variant file though a
preprocessor macro.

This allows to keep the existing build system intact, but also redefine
at compile time the final header without having to make copies on the
file system.

wavexx committed:
cmake: Define FW_VARIANT instead of copying Configuration_prusa

wavexx commented:
I went with the above approach. Not only it avoids the need to copy/check Configuration_prusa, but you have the added advantage that you get compiler errors pointing _correct file_ when changing the variant header.

wavexx committed:
cmake: Remove unneeded additional include path

3d-gussner reviewed (approved):
Tested on MK404
Compiling with `PF-build.sh` and `cmake` works even with `Configuration_prusa.h` existing.

wavexx commented:
Merging then
